aten/src/ATen/native/ComplexHelper.cpp

```diff
@@ -0,0 +1,54 @@
+#include <ATen/native/ComplexHelper.h>
+
+#include <ATen/native/TensorIterator.h>
+
+namespace at {
+namespace native {
+
+DEFINE_DISPATCH(complex_stub);
+DEFINE_DISPATCH(complex_polar_stub);
+
+Tensor& complex_out(Tensor& result, const Tensor& real, const Tensor& imag) {
+  TORCH_CHECK(real.scalar_type() == imag.scalar_type(),
+              "Expected object of scalar type ", real.scalar_type(), " but got "
+              "scalar type ", imag.scalar_type(), " for argument 'imag'");
+  // Sort of hacky, but necessary so that input dtypes don't get promoted.
+  auto iter = TensorIterator::comparison_op(result, real, imag,
+    /*check_mem_overlap=*/true);
+  complex_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor complex(const Tensor& real, const Tensor& imag) {
+  c10::TensorOptions options = real.options();
+  switch (real.scalar_type()) {
```

Comment:
I don't think this is correct. we should instead do:
```
switch (promoteTypes(t.scalar_type(), src.scalar_type()) {
   case c10::kDouble: options = options.dtype(c10::kComplexDouble); break;
   default:
      options.dtype(c10::kComplexFloat); break;
};

auto result = at::empty(real.sizes(), options);
```

Reply:
sounds good!